### ISSUE TITLE
qml.jacobian returns 0 for all circuits

### ISSUE THREAD

**Issue by ijdavis** [PennyLaneAI/pennylane#4835]
### Expected behavior

```python
def circuit_function(weights):
    qml.Hadamard(wires=0)
    qml.RY(weights[0], wires=0)
    return [ qml.expval(qml.PauliZ(wires=0)) ]

weights         = [0]
control_weights = np.array(weights, requires_grad=True)

dev   = qml.device("default.qubit", wires=1, shots=None)
qNode = qml.QNode(circuit_function, dev, diff_method="backprop")

gradients   = qml.jacobian(lambda x: np.stack(qNode(x)))(control_weights)
slope       = sum(gradients)
print(f'slope={slope}')
print(f'gradients={gradients}')
```

-------------------------------
Expected behavour is for gradients to return

```python
slope=[-1.]
gradients=[[-1.]]
```

### Actual behavior

```python
slope=[0]
gradients=[[0]]
```

change weights to [0.0] fixed this problem

```python
slope=[-1.]
gradients=[[-1.]]
```

### Additional information

This behaviour was really hard to track down and could easily break code.  PennyLane code should be assuming that rotations are floating point values even if presented in the control weights as integer values.

Thanks,
Ian
cs.uwaterloo.ca/~ijdavis

### Source code

```shell
def circuit_function(weights):
    qml.Hadamard(wires=0)
    qml.RY(weights[0], wires=0)
    return [ qml.expval(qml.PauliZ(wires=0)) ]

weights         = [0]
control_weights = np.array(weights, requires_grad=True)

dev   = qml.device("default.qubit", wires=1, shots=None)
qNode = qml.QNode(circuit_function, dev, diff_method="backprop")

gradients   = qml.jacobian(lambda x: np.stack(qNode(x)))(control_weights)
slope       = sum(gradients)
print(f'slope={slope}')
print(f'gradients={gradients}')
```

### System information

```python
Name: PennyLane
Version: 0.33.1
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane-Lightning

Platform info:           Linux-5.15.120+-x86_64-with-glibc2.35
Python version:          3.10.12
Numpy version:           1.23.5
Scipy version:           1.11.3
Installed devices:
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by isaacdevlugt** [comment]
Hey @ijdavis! Looks like this is being worked on in the PR above 👆. Once that gets merged and your issue is resolved, you can access the bug fix by installing pennylane from source.

**Comment by ijdavis** [comment]
Attention: Isaac De Vlugt..

I am writing a QIC890 term paper at University of Waterloo (which may be
publishable) on research which leveraged PennyLane.

Any concerns or improvement regarding including this in my acknowledgement??

Software was developed using Google Colab and  Xanada PennyLane. Isaac De
Vlugt (PennyLane Team) was friendly, supportive and responsive to all
software concerns raised.

Regards, Ian
cs.uwaterloo.ca/~ijdavis

**Comment by ijdavis** [comment]
A question if I may :

I am implementing Quantum State Tomography.

It creates hundreds of gates and it is incredibly inefficient to ask for
expectation values on circuits containing ever more gates (4 hours +).
Initially obtaining an expectation value takes seconds.. by the end it is
10 minutes or more per expectation value.

I would love to be able to say after each small step, permanently change
the device so its new initial quantum state is the one that results from
applying these prior gates.
Then I could throw gates away after applying them and apply very short gate
sequences to each permanently changed quantum state in the device at each
step.
I want to incrementally improve the quantum state permanently , not
incrementally add ever more gates to my circuit which must all be
repeatedly executed.

By analogy

list = []
for i in range(1e10):
    list.append(i)
    sum1 = sum(list)
    use(sum1)

Is incredibly slow compared to:

sum1 = 0
for i in range(1e10)
    sum1 += i
    use(sum1)



================================================

It looks like Qnode.construct() might do this.  But there is no
documentation on Qnode.construct beyond it needing arguments and keywords.

I've got as far as (guessing):
devExpval   = qml.device("default.qubit", wires=WIRES, shots=None)
qNode = qml.QNode(function, devExpval, diff_method="backprop")
# Construct the tape
qNode(control_weights)
# Recover the tape
tape = qNode.qtape
print(f'qtape={tape}')

qtape=<QuantumScript: wires=[0, 1, 2, 3, 4, 5, 6], params=14>
#qml.execute((tape,), devExpval)
devExpval.execute(tape)

Would this permanently change the device's initial state???
What about a device which returns shots.. There are comments about not
being able to do this on such devices??

If this should work, why doesn't it work? Am I looking at a bug below??

Or am I currently basically trying to do the impossible??

Thanks,

Ian
cs.uwaterloo.ca/~ijdavis

 421                     devExpval.execute(tape)

/usr/local/lib/python3.10/dist-packages/pennylane/devices/default_qubit.py
<https://localhost:8080/#> in execute(self, circuits,
execution_config)    472         )    473         if max_workers is
None:--> 474             results = tuple(    475
simulate(    476                     c,

/usr/local/lib/python3.10/dist-packages/pennylane/devices/default_qubit.py
<https://localhost:8080/#> in <genexpr>(.0)    473         if
max_workers is None:    474             results = tuple(--> 475
         simulate(    476                     c,    477
     rng=self._rng,

/usr/local/lib/python3.10/dist-packages/pennylane/devices/qubit/simulate.py
<https://localhost:8080/#> in simulate(circuit, rng, prng_key,
debugger, interface)    267     268     """--> 269     state,
is_state_batched = get_final_state(circuit, debugger=debugger,
interface=interface)    270     return measure_final_state(circuit,
state, is_state_batched, rng=rng, prng_key=prng_key)

/usr/local/lib/python3.10/dist-packages/pennylane/devices/qubit/simulate.py
<https://localhost:8080/#> in get_final_state(circuit, debugger,
interface)    159     is_state_batched = bool(prep and prep.batch_size
is not None)    160     for op in circuit.operations[bool(prep) :]:-->
161         state = apply_operation(op, state,
is_state_batched=is_state_batched, debugger=debugger)    162     163
      # Handle postselection on mid-circuit measurements

/usr/lib/python3.10/functools.py <https://localhost:8080/#> in
wrapper(*args, **kw)    887                             '1 positional
argument')    888 --> 889         return
dispatch(args[0].__class__)(*args, **kw)    890     891     funcname =
getattr(func, '__name__', 'singledispatch function')

/usr/local/lib/python3.10/dist-packages/pennylane/devices/qubit/apply_operation.py
<https://localhost:8080/#> in apply_operation(op, state,
is_state_batched, debugger)    196     197     """--> 198     return
_apply_operation_default(op, state, is_state_batched, debugger)    199
    200

/usr/local/lib/python3.10/dist-packages/pennylane/devices/qubit/apply_operation.py
<https://localhost:8080/#> in _apply_operation_default(op, state,
is_state_batched, debugger)    206         and math.ndim(state) <
EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD    207     ) or (op.batch_size
and is_state_batched):--> 208         return
apply_operation_einsum(op, state, is_state_batched=is_state_batched)
 209     return apply_operation_tensordot(op, state,
is_state_batched=is_state_batched)    210

/usr/local/lib/python3.10/dist-packages/pennylane/devices/qubit/apply_operation.py
<https://localhost:8080/#> in apply_operation_einsum(op, state,
is_state_batched)     70         array[complex]: output_state     71
  """---> 72     mat = op.matrix()     73      74     total_indices =
len(state.shape) - is_state_batched

/usr/local/lib/python3.10/dist-packages/pennylane/operation.py
<https://localhost:8080/#> in matrix(self, wire_order)    773
   tensor_like: matrix representation    774         """--> 775
 canonical_matrix = self.compute_matrix(*self.parameters,
**self.hyperparameters)    776     777         if wire_order is None
or self.wires == Wires(wire_order):

/usr/local/lib/python3.10/dist-packages/pennylane/operation.py
<https://localhost:8080/#> in compute_matrix(*params, **hyperparams)
 742             tensor_like: matrix representation    743
"""--> 744         raise MatrixUndefinedError    745     746     #
pylint: disable=no-self-argument, comparison-with-callable

MatrixUndefinedError:


Thanks,

Ian Davis
cs.uwaterloo.ca/~ijdavis

**Comment by albi3ro** [comment]
What I might recommend is returning both the measurement that you want and the state via `qml.state()`. Then you can initialize your next circuit into that value with `qml.QubiStateVector`.

Something like:
```
@qml.qnode(dev)
def circuit(initial_state):
    qml.QubitStateVector(initial_state, wires=(0,))
    # additional gates
    return qml.state(), qml.expval(qml.PauliZ(0))
```

**Comment by ijdavis** [comment]
P.S.  Hope this warms your heart.... code implemented as you suggested...

Your excellent advice reduced my run time from 15403 seconds (4.27 hours)
to 1185 seconds (19.75 minutes).

That is a huge performance improvement (albeit not practical on a real
quantum computer).

Thanks for the awesome suggestion..

Regards, Ian
cs.uwaterloo.ca/~ijdavis
